piker/brokers/binance.py

```diff
@@ -245,33 +303,76 @@ async def bars(
             }
         )
 
-        # TODO: pack this bars scheme into a ``pydantic`` validator type:
-        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
-
-        # TODO: we should port this to ``pydantic`` to avoid doing
-        # manual validation ourselves..
         new_bars = []
         for i, bar in enumerate(bars):
+            bar = KLine(**{
+                key: bar[j]
+                for j, key in enumerate(
+                    KLine.__dict__['__fields__'].keys())
+            })
+            bar_wap = .0
+            new_bars.append((i,) + tuple(bar.as_row()) + (bar_wap,))
+
+        array = np.array(
+            new_bars,
+            dtype=ohlc_dtype
+        ) if as_np else bars
+        
+        return array
 
-            bar = OHLC(*bar)
-
-            row = []
-            for j, (name, ftype) in enumerate(_ohlc_dtype[1:]):
-
-                # TODO: maybe we should go nanoseconds on all
-                # history time stamps?
-                if name == 'time':
-                    # convert to epoch seconds: float
-                    row.append(bar.time / 1000.0)
-
-                else:
-                    row.append(getattr(bar, name))
+    async def submit_limit(
+        self,
+        symbol: str,
+        side: str,  # SELL / BUY
+        quantity: float,
+        price: float,
+        # time_in_force: str = 'GTC',
+        oid: Optional[int] = None,
+        # iceberg_quantity: Optional[float] = None,
+        # order_resp_type: Optional[str] = None,
+        recv_window: int = 60000
+    ) -> int:
+        symbol = symbol.upper()
+
+        await self.cache_symbols()
+
+        asset_precision = self._pairs[symbol]['baseAssetPrecision']
+        quote_precision = self._pairs[symbol]['quoteAssetPrecision']
+
+        quantity = Decimal(quantity).quantize(
```

Comment:
We need to sanitize `price`/`quantity` because `binance` has specific per-symbol requirements on precision.

Reply:
Yeah this is also going to play in with new $positioning features where a user can specify the base pair for the position size; which fits naturally with crypto / forex, less so with discrete instruments like stocks / most derivatives.